parachain/pallets/bridge/src/lib.rs

```diff
@@ -1,56 +1,73 @@
 #![cfg_attr(not(feature = "std"), no_std)]
+#![allow(unused_variables)]
 
-use frame_support::{decl_module, decl_storage, decl_event, decl_error, dispatch};
+use frame_support::{decl_error, decl_event, decl_module, decl_storage, dispatch::DispatchResult};
 use frame_system::{self as system, ensure_signed};
 
 use sp_std::prelude::*;
 
-use sp_runtime::traits::Hash;
-
-use artemis_core::{AppID, Message, Broker};
+use artemis_core::{
+	registry::{App, lookup_app},
+	AppId, Application, Message, Verifier,
+};
 
 pub trait Trait: system::Trait {
-
-	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
-
-	type Broker: Broker;
+	type Event: From<Event> + Into<<Self as system::Trait>::Event>;
+	type Verifier: Verifier<<Self as system::Trait>::AccountId>;
+	type AppETH: Application;
+	type AppERC20: Application;
 }
 
 decl_storage! {
 	trait Store for Module<T: Trait> as BridgeModule {
+
 	}
 }
 
 decl_event!(
-	pub enum Event<T>
-	where
-		AccountId = <T as system::Trait>::AccountId,
-		Hash = <T as frame_system::Trait>::Hash,
-	{
-		Received(AccountId, AppID, Hash),
+	pub enum Event {
 	}
 );
 
 decl_error! {
 	pub enum Error for Module<T: Trait> {
+		HandlerNotFound
 	}
 }
 
 decl_module! {
-
 	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
 
 		type Error = Error<T>;
 
 		fn deposit_event() = default;
 
+		/// Submit `message` for dispatch to an application identified by `app_id`.
 		#[weight = 0]
-		pub fn send(origin, app_id: AppID, message: Message) -> dispatch::DispatchResult {
+		pub fn submit(origin, app_id: AppId, message: Message) -> DispatchResult {
 			let who = ensure_signed(origin)?;
-			T::Broker::submit(app_id, message.clone())?;
+			let app = lookup_app(app_id).ok_or(Error::<T>::HandlerNotFound)?;
+			Self::verify(who, app_id, &message)?;
+			Self::dispatch(app, message)
+		}
+
+	}
+}
 
-			Self::deposit_event(RawEvent::Received(who, app_id, T::Hashing::hash(message.as_ref())));
-			Ok(())
+impl<T: Trait> Module<T> {
+
+	// verify message
+	fn verify(sender: T::AccountId, app_id: AppId, message: &Message) -> DispatchResult {
+		T::Verifier::verify(sender, app_id, &message)
+	}
+
+	/// Dispatch `message` to application `app`.
+	fn dispatch(app: App, message: Message) -> DispatchResult {
+		match app {
+			App::ETH => T::AppETH::handle(message),
+			App::ERC20 => T::AppERC20::handle(message)
 		}
```

Comment:
Would be difficult to achieve given constraints with how Substrate implements pallets and how it loosely couples them together via traits.

Firstly, as a primer, the `Application` type is a trait that is implemented by `pallet_eth::Module` and `pallet_erc20::Module` pallet modules. These modules are never instantiated. Indeed Substrate doesn't instantiate any of its pallets. It just calls static methods on the type.

That means a hypothetical registry lookup function would never be able to return a instantiated application _object_. And Rust doesn't support first-class types (i.e passing types as parameters into functions, or returning types from functions).

That limitation forces us to implement the hardcoded message dispatch in the bridge module, because that's the first place where `Application` types were injected by the runtime:
https://github.com/Snowfork/polkadot-ethereum/blob/main/parachain/runtime/src/lib.rs#L277

Reply:
Understood – is there another way around concrete types in the bridge though? You could just pass the registry itself as a type to the bridge pallet, and the registry gets filled with applications upon its instantiation? Then `handle` would be called on the registry, which determines the right application at runtime?